reconcile/vault_replication.py

```diff
@@ -0,0 +1,252 @@
+            raise VaultInvalidPaths
+
+
+def list_invalid_paths(path_list: List[str], policy_paths: List[str]) -> List[str]:
+    invalid_paths = []
+
+    for path in path_list:
+        if not policy_contains_path(path, policy_paths):
+            invalid_paths.append(path)
+
+    return invalid_paths
+
+
+def policy_contains_path(path: str, policy_paths: List[str]) -> bool:
+    return any(path in p_path for p_path in policy_paths)
+
+
+def get_policy_paths(policy_name, instance_name) -> List[str]:
+    query_data = vault_policies.query(query_func=gql.get_api().query)
+    policy_paths = []
+
+    if query_data.policy:
+        for policy in query_data.policy:
+            if policy.name == policy_name and policy.instance.name == instance_name:
+                for line in policy.rules.split("\n"):
+                    res = re.search(r"path \s*[\'\"](.+)[\'\"]", line)
+
+                    if res is not None:
+                        policy_paths.append(res.group(1))
+
+    return policy_paths
+
+
+def get_jenkins_secret_list(jenkins_instance: str) -> List[str]:
+    """Returns a list of secrets used in a jenkins instance
+
+    The input secret is the name of a jenkins instance to filter
+    the secrets:
+    * jenkins_instance - Jenkins instance name
+    """
+    secret_list = []
+    query_data = jenkins_configs.query(query_func=gql.get_api().query)
+
+    if query_data.jenkins_configs:
+        for p in query_data.jenkins_configs:
+            if (
+                isinstance(p, JenkinsConfigV1_JenkinsConfigV1)
+                and p.instance.name == jenkins_instance
+                and p.config_path
+            ):
+                secret_paths = [
+                    line
+                    for line in p.config_path.content.split("\n")
+                    if "secret-path" in line
+                ]
+                for line in secret_paths:
+                    res = re.search(r"secret-path:\s*[\'\"](.+)[\'\"]", line)
+                    if res is not None:
+                        secret_list.append(res.group(1))
+
+    return secret_list
+
+
+def get_vault_credentials(
+    vault_instance: Union[VaultInstanceV1, VaultReplicationConfigV1_VaultInstanceV1]
+) -> dict[str, Optional[str]]:
+    """Returns a dictionary with the credentials used to authenticate with Vault,
+    retrieved from the values present on AppInterface.
+
+    The input is a VaultInstanceV1 object, that contains secret references to be retreived
+    from vault. The output is a dictionary with the credentials used to authenticate with
+    Vault.
+    * vault_instance - VaultInstanceV1 object from AppInterface Data.
+    """
+    vault_creds = {}
+    vault = cast(_VaultClient, VaultClient())
+
+    if not isinstance(
+        vault_instance.auth, VaultInstanceAuthApproleV1
+    ) and not isinstance(
+        vault_instance.auth,
+        VaultReplicationConfigV1_VaultInstanceV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
```

Comment:
mypy complaining again? 😢

Reply:
Since I removed the cast to `VaultInstanceV1` the object can be two different things, the `VaultInstanceV1` itself, or the VaultInstance defined inside the replication part of the schema. That second one is the type of `VaultReplicationConfigV1_VaultInstanceV1` and mypy / the isinstance was not happy

Reply:
ok!